### PR TITLE
Websocket improvements

### PR DESCRIPTION
**Pull request by toini**

### CODE REVIEW

File: drivers/pulse/device.ts [hunk 1]
```diff
@@ -104,21 +110,27 @@ class PulseDevice extends Device {
 
   async #subscribeToLive() {
     this.#resubscribeDebounce();
+
+    let {
```
Comment (jstaro) [inline drivers/pulse/device.ts:114]:
```suggestion
    const {
```

File: drivers/watty/device.ts [hunk 2]
```diff
@@ -69,10 +70,15 @@ class WattyDevice extends Device {
       })`,
     );
 
-    // Resubscribe if no data for 10 minutes
+    // Jitter to avoid all clients resubscribing at same time after API reboot
+    const jitter = getRandomDelay(0, 10);
+    const delay = 10 * 60; // Ten minutes in seconds
+    this.#resubscribeMaxWait = (jitter + delay) * 1000; // Milliseconds
```
Comment (jstaro) [inline drivers/watty/device.ts:76]: Naming: same deal as in `pulse/device.ts`

File: drivers/watty/device.ts [hunk 3]
```diff
@@ -103,20 +109,27 @@ class WattyDevice extends Device {
 
   async #subscribeToLive() {
     this.#resubscribeDebounce();
+
+    let {
```
Comment (jstaro) [inline drivers/watty/device.ts:113]:
```suggestion
    const {
```

File: lib/tibber.ts [hunk 4]
```diff
@@ -341,28 +342,37 @@ export class TibberApi {
...
       },
-      webSocketImpl: ws,
+      webSocketImpl: UserAgentWebSocket,
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      // shouldRetry: (errOrCloseEvent: any) => {
+      //   this.#log(
+      //     `Error while subscribing to live; code ${errOrCloseEvent.code}; reason ${errOrCloseEvent.reason}`,
+      //   );
+      //   return errOrCloseEvent !== CloseCode.Forbidden;
+      // },
```
Comment (jstaro) [inline lib/tibber.ts:363]: Remove.

**Review by jstaro — CHANGES_REQUESTED** [review]
Looks good! Just some tiny naming suggestions and removing some commented-out code.

File: lib/tibber.ts (hunk 4 above)
Comment (toini) [reply lib/tibber.ts:363]: Im still testing this part. Turned out atm server doesnt close the connection when it shutsdown. I'll try to make it do that (node18 even has a new function for closing all http server connections). I believe we should arrive in this method when server closes with fatal http status.

File: drivers/pulse/device.ts (hunk 1 above)
Comment (toini) [reply drivers/pulse/device.ts:114]: But websocketSubscriptionUrl needs to be a variable as it gets updated, right?

File: lib/tibber.ts (hunk 4 above)
Comment (toini) [reply lib/tibber.ts:363]: No need for node18. serverCleanup.dispose() does the trick. Just needed to add a temporary endpoint for calling that (when running in debugger it seems to die without any dispose - would need to test this on server too).

File: drivers/watty/device.ts (hunk 3 above)
Comment (toini) [reply drivers/watty/device.ts:113]: websocketSubscriptionUrl needs to be variable.

File: drivers/pulse/device.ts (hunk 1 above)
Comment (jstaro) [reply drivers/pulse/device.ts:114]: Ah, right. Couldn't compile that code in my head, sorry.

**Pull request merged**